File: packages/typescriptlib/src/test/typescript/serviceInstanceProvider.test.ts
import {BaseService} from "@src/typescript/service";
import {DefaultServiceInstanceProviderImpl} from "@src/typescript/serviceInstanceProvider";
import {describe, expect, test} from "vitest";

// As I don't know how to properly spy on a constructor, I'm creating a simple service that can indicate if the
// constructor was called
class TestService extends BaseService
{
    wasCalled: boolean = false;
    constructor()
    {
        super();
        this.wasCalled = true;
    }
};

describe("Service Instance Provider", () => {
    test("DefaultServiceInstanceProviderImpl should create a new instance of the service", () => {
        const provider = new DefaultServiceInstanceProviderImpl(TestService);
        const instance = provider.createServiceInstance();

        expect(instance.wasCalled).toBe(true);
        expect(instance instanceof TestService).toBe(true);
    });
});

File: packages/pythonlib/src/test/python/sphyrna/service_manager/service_instance_provider_test.py
import pytest

from sphyrna.service_manager.service import BaseService
from sphyrna.service_manager.service_instance_provider import (
    DefaultServiceInstanceProviderImpl,
)


class MockService(BaseService):
    def __init__(self):
        self.was_called = True


def test_constructor_and_create_service_instance():
    # Create an instance of DefaultServiceInstanceProviderImpl with TestService as a parameter
    service_instance_provider = DefaultServiceInstanceProviderImpl(MockService)

    # Call createServiceInstance on the DefaultServiceInstanceProviderImpl instance
    service_instance = service_instance_provider.create_service_instance()

    # Assert that the constructor of TestService was called
    assert service_instance.was_called

    # Assert that the service instance is an instance of TestService
    assert isinstance(service_instance, MockService)
